File: 009-single-number-iii/Yiqi_Ou.java
public class Solution {
    public int[] singleNumber(int[] nums) {
        int[] res = new int[2];
        if(nums == null || nums.length < 2)
            return res;
        int xor = 0;
        for(int i : nums)  
            xor = xor ^ i;
        int mask = 1;
        while((xor & mask) == 0)
            mask = mask << 1;
        int xor1 = 0, xor2 = 0;
        for(int i : nums) {
            if((mask & i) == 0) {
                xor1 = xor1 ^ i;
            }
            else {
                xor2 = xor2 ^ i;
            }
        }
        res[0] = xor1;
        res[1] = xor2;
        return res;
    }
}

File: 009-single-number-iii/Yiqi_Ou.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    var xor = 0;
    nums.forEach(function(item, idx, array) {
       xor ^= item; 
    });
    var mask = 1;
    while((xor & mask) === 0)
        mask <<= 1;
    var res = [];
    var num1 = 0, num2 = 0;
    nums.forEach(function(item, idx, array) {
        if(item & mask)
            num1 ^= item;
        else 
            num2 ^= item;
    });
    res.push(num1, num2);
    return res;
};